feat: Add features auto-completion

BertrandRenault opened the pull request:
close #76 

Notes: 
- for Scene features, as there is often/always a default value, the auto-completion only show the current option, others options are shown if we delete current value
![image](https://github.com/pixano/pixano/assets/112549869/23390b73-8342-4c62-be2f-ff173e25ce2c)
![image](https://github.com/pixano/pixano/assets/112549869/093a93fc-b147-4c7b-8381-1b6405282620)

- available values are read from tables in back, not from db.json. This could be done in a future PR, if necessary
- (back) merge with "categories" -> related to previous, if we choose to save available values in db.json, we will refactor "categories" to be more generic

timothee-LJN commented on ui/components/core/src/lib/types/datasetTypes.ts:
```diff
@@ -139,6 +140,19 @@ export interface ItemFeature {
 
+export interface FeaturesAvailableValues {
+  [key: string]: {
+    [innerKey: string]: Array<string>;
```
your index here is not a variable, I think its more idiomatic to call always the same 

```

export interface FeaturesAvailableValues {
  [key: string]: {
    [key: string]: Array<string>;
  };
}
```

timothee-LJN commented on ui/components/imageWorkspace/src/ImageWorkspace.svelte:
```diff
@@ -104,6 +109,30 @@
+        }
+        itemFeaturesAvailableValues.set(outputFeaturesAvailableValues);
+      })
+      .catch((err) => console.log(err));
```
maybe used `console.error(err)` here?

timothee-LJN commented on ui/components/imageWorkspace/src/ImageWorkspace.svelte:
```diff
@@ -104,6 +109,30 @@
+          datasetWithFeats.available_feat_values || {};
+        let outputFeaturesAvailableValues: FeaturesAvailableValues = {};
+        //convert from InputFeaturesAvailableValues to FeaturesAvailableValues
+        for (let table in inputFeaturesAvailableValues) {
```
Following our conversation, you could have used a reduce like so : 
```
        const outputFeaturesAvailableValues: FeaturesAvailableValues = Object.entries(
          inputFeaturesAvailableValues,
        ).reduce((tables, [tableName, tableValues]) => {
          tables[tableName] = tableValues.reduce(
            (features, feature) => {
              if (feature.values) {
                features[feature.name] = feature.values;
              }
              return features;
            },
            {} as FeaturesAvailableValues[""],
          );
          return tables;
        }, {} as FeaturesAvailableValues);
```

I personnaly tend to prefer reduce but a for in loop works also well!

timothee-LJN commented on ui/components/imageWorkspace/src/components/Features/FeatureFormInputs.svelte:
```diff
@@ -99,13 +99,21 @@
         step={feature.type === "int" ? "1" : "any"}
         autofocus={i === 0 ? true : false}
         value={initialValues[feature.name]?.value || ""}
+        list="obj_availableValues_{feature.name}"
```
use lowerCamelCase !

timothee-LJN commented on ui/components/imageWorkspace/src/components/Features/FeatureFormInputs.svelte:
```diff
@@ -99,13 +99,21 @@
             feature.name,
           )}
       />
+      <datalist id="obj_availableValues_{feature.name}">
```
use lowerCamelCase !

timothee-LJN commented on ui/components/imageWorkspace/src/components/Features/TextInputFeature.svelte:
```diff
@@ -43,12 +55,21 @@
-      on:input={() => (isSaved = false)}
       type={inputType === "str" ? "text" : "number"}
       step={inputType === "int" ? "1" : "any"}
+      list="scene_availableValues_{textFeature.name}"
```
use lowerCamalCase here!

timothee-LJN commented on ui/components/imageWorkspace/src/components/Features/TextInputFeature.svelte:
```diff
@@ -43,12 +55,21 @@
       on:keyup={(e) => e.stopPropagation()}
     />
+    <datalist id="scene_availableValues_{textFeature.name}">
+      {#if "scene" in $itemFeaturesAvailableValues && textFeature.name in $itemFeaturesAvailableValues.scene}
```
I think we could narrow down `FeaturesAvailableValues` to something more explicitif 

```
interface FeaturesAvailableValues {
  scene: Record<string, Array<string>>;
  object: Record<string, Array<string>>;
}
```

that would make your component type safer and would avoid these type check everywhere. That would also make the code more readable. 
If the future we add more categories we will either refacto to something more generic or add the new category to the type.

cpvannier reviewed (changes requested):
Regarding the feature itself, currently this PR is only working for me on scene features. On object features, it only suggests me the current value after I hit enter, not the existing values.

Regarding the content of the PR, I think the TMP code in `dataset.py` and `ImageWorkspace.svelte` should be fixed here instead of in a future PR. I consider this feature to be a nice-to-have rather than a necessity, so I don't think we need to rush merging this into develop with temporary solutions.

The `categories` attribute of `DatasetInfo` should also be removed, along with its `DatasetCategory` class, as it is deprecated by `available_feat_values` and no longer needed.

The conversion from `InputFeaturesAvailableValues` to `FeaturesAvailableValues` in the frontend is a workaround that should be fixed as well. One solution could be to remove the `name` attribute from `DatasetAvailableFeatureValues` in the backend, and change the type of `available_feat_values` attribute in `DatasetInfo` to `dict[str, dict[str, DatasetAvailableFeatureValues]]` so that the feature names are already accessible as dictionary keys in the backend.

I also think the class and variable names are too long, I would rename to something like this:
- Backend
  - Rename `DatasetAvailableFeatureValues` class to `FeatureValues` and move it inside pixano/data/item/item_feature.py
  - Rename `available_feat_values` attribute of `DatasetInfo` to `feature_values` 
  - Rename `load_available_values` argument of `Dataset.load_info()` to `load_feature_values`
- Frontend
  - Remove `InputFeaturesAvailableValues` interface as mentioned above (have the correct format in the backend already)
  - Rename `FeaturesAvailableValues` interface to `FeatureValues`
  - Remove `inputFeaturesAvailableValues` and `itemFeaturesAvailableValues` variables as mentioned above (have `featureValues` in `DatasetInfo` instead)

timothee-LJN commented on ui/components/imageWorkspace/src/components/SaveShape/SaveShapeForm.svelte:
```diff
@@ -76,6 +81,21 @@
     });
+
+    // add new inputs to lists of available values
+    if ("objects" in $itemFeaturesAvailableValues) {
```
this function looks the same as in `TextInputFeature.svelte`. Maybe you could abstract it in a function in an api file ?

timothee-LJN commented on ui/components/imageWorkspace/src/components/Features/TextInputFeature.svelte:
```diff
@@ -43,12 +55,21 @@
+      on:input={() => (isSaved = false)}
       on:keyup={(e) => e.stopPropagation()}
     />
+    <datalist id="scene_availableValues_{textFeature.name}">
```
there is a `Combobox` component available (only available with search feature for now but could be easily improved). 
You can find documentation here : https://www.shadcn-svelte.com/docs/components/combobox 

I can help on that if you want?

timothee-LJN commented on ui/components/imageWorkspace/src/components/Features/TextInputFeature.svelte:
```diff
@@ -43,12 +55,21 @@
     />
+    <datalist id="scene_availableValues_{textFeature.name}">
+      {#if "scene" in $itemFeaturesAvailableValues && textFeature.name in $itemFeaturesAvailableValues.scene}
+        {#each $itemFeaturesAvailableValues.scene[textFeature.name].sort() as proposedValue}
```
that's cool to use reactive store, i never do it but will more in the future!

timothee-LJN commented on ui/components/imageWorkspace/src/ImageWorkspace.svelte:
```diff
@@ -104,6 +109,30 @@
+    // later, we should get itemFeaturesAvailableValues directly from DatasetInfo["available_feat_values"]
+    api
+      .getDataset(currentDatasetId)
+      .then((datasetWithFeats) => {
```
I feel like this is not located in the most accurate component. Data fetching is mainly done in `src/routes/layout.svelte` . Maybe that could be something to reconsider but for now I think we should try to locate as many connexion to the API with in there, what do you think?

BertrandRenault replied:
Done

BertrandRenault commented:
I didn't remove "categories" as proposed yet, because without we can't export directly to COCO format.
Need to statute about the change request first

BertrandRenault commented:
> Regarding the feature itself, currently this PR is only working for me on scene features. On object features, it only suggests me the current value after I hit enter, not the existing values.

Auto-completion for object edition was omitted. Now added.

> Regarding the content of the PR, I think the TMP code in `dataset.py` and `ImageWorkspace.svelte` should be fixed here instead of in a future PR. I consider this feature to be a nice-to-have rather than a necessity, so I don't think we need to rush merging this into develop with temporary solutions.
>
> The `categories` attribute of `DatasetInfo` should also be removed, along with its `DatasetCategory` class, as it is deprecated by `available_feat_values` and no longer needed.

 
> The conversion from `InputFeaturesAvailableValues` to `FeaturesAvailableValues` in the frontend is a workaround that should be fixed as well. One solution could be to remove the `name` attribute from `DatasetAvailableFeatureValues` in the backend, and change the type of `available_feat_values` attribute in `DatasetInfo` to `dict[str, dict[str, DatasetAvailableFeatureValues]]` so that the feature names are already accessible as dictionary keys in the backend.

 
> I also think the class and variable names are too long, I would rename to something like this:
> 
> * Backend
>   
>   * Rename `DatasetAvailableFeatureValues` class to `FeatureValues` and move it inside pixano/data/item/item_feature.py
class removed, use of primitive type instead (dict[str, dict[str, list[string]]])

>   * Rename `available_feat_values` attribute of `DatasetInfo` to `feature_values`
Done

>   * Rename `load_available_values` argument of `Dataset.load_info()` to `load_feature_values`
Renamed to 'load_features_values'

> * Frontend
>   
>   * Remove `InputFeaturesAvailableValues` interface as mentioned above (have the correct format in the backend already)
Done

>   * Rename `FeaturesAvailableValues` interface to `FeatureValues`
Done
>   * Remove `inputFeaturesAvailableValues` and `itemFeaturesAvailableValues` variables as mentioned above (have `featureValues` in `DatasetInfo` instead)
Not yet

cpvannier commented:
> I didn't remove "categories" as proposed yet, because without we can't export directly to COCO format. Need to statute about the change request first

True, this deserves its separate issue: pixano/pixano-project-manager#46

Also, regarding storing `feature_values` versus fetching the data with SQL requests, Jaonary suggested merging this PR as is with SQL requests only, and to look into adding a backend cache for storing data like this while the app is open, rather that writing it in the `db.json` files: pixano/pixano-project-manager#45
